apis/v1alpha1/validation.go

```diff
@@ -33,11 +36,71 @@ func ValidateResourceMapping(m *ResourceMapping) (vErr error) {
 			vErr = errors.Join(vErr, fmt.Errorf("invalid owner: %w", err))
 		}
 	}
-	if m.Resource.Provider == "" {
-		vErr = errors.Join(vErr, fmt.Errorf("empty resource provider"))
-	}
+
 	if _, ok := m.Annotations.AsMap()[AnnotationKeyAssetInfo]; ok {
 		vErr = errors.Join(vErr, fmt.Errorf("reserved key is included: %s", AnnotationKeyAssetInfo))
 	}
+
+	if err := validateResource(m.Resource); err != nil {
+		vErr = errors.Join(vErr, fmt.Errorf("invalid Resource: %w", err))
+	}
+
+	return
+}
+
+func validateResource(r *Resource) (vErr error) {
+	if r.Name == "" {
+		vErr = errors.Join(vErr, fmt.Errorf("empty resource name"))
+	}
+
+	if r.Provider == "" {
+		vErr = errors.Join(vErr, fmt.Errorf("empty resource provider"))
+	}
+
+	if err := validateSubscope(r); err != nil {
+		vErr = errors.Join(vErr, fmt.Errorf("invalid subscope: %w", err))
+	}
+
 	return
 }
+
+func validateSubscope(r *Resource) error {
+	if r.Subscope == "" {
+		return nil
+	}
+
+	// If r.Subscope = "parent/foo/child/bar?key1=value1&key2=value2"
+	// after url.Parse(r.Subscope), we will have:
+	// u.RawQuery: key1=value1&key2=value2
+	u, err := url.Parse(r.Subscope)
+	if err != nil {
+		return fmt.Errorf("failed to parse subscope string: %w", err)
+	}
+
+	// ParseQuery is needed here to check if query string has errors
+	q, err := url.ParseQuery(u.RawQuery)
+	if err != nil {
+		return fmt.Errorf("failed to parse qualifier string: %w", err)
+	}
+
+	keys := make([]string, 0, len(q))
+	for k := range q {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	var kvPairs []string
+	for _, k := range keys {
+		sort.Strings(q[k])
```

Comment:
Why do you need to sort here? I thought the key should be in order not the values

Reply:
Yes, otherwise if a key has multiple values, it would cause problem. For example: `key1=value1&key1=value2&key2=value3` and `key1=value2&key1=value1&key2=value3`